modules/validations/ambiguous_crossing_tags.js

```diff
@@ -84,174 +159,272 @@ export function validationAmbiguousCrossingTags(context) {
-              const node = graph.hasEntity(nodeID);
-              const way = graph.hasEntity(wayID);
-              if (!node || !way) return;
-
-              const tags = Object.assign({}, node.tags);
-              Object.keys(tags).forEach(tag => {
-                if ((tag === 'crossing') || (tag === 'crossing:markings')) return;
-                delete tags.tag;
-              });
-
-              const annotation = l10n.t('issues.fix.use_crossing_tags_from_node.annotation');
-              editor.perform(actionChangeTags(wayID, tags));
+            icon: 'rapid-icon-crossing',
+            title: getTitle(parentWay.tags),
+            onClick: () => {
+              const annotation = l10n.t(
+                'issues.fix.set_both_as_marked.annotation'
+              );
+              editor.perform(doMarkBothAsWay, annotation);
               editor.commit({ annotation: annotation, selectedIDs: context.selectedIDs() });
             }
-          })
+          },
+          )
         );
       }
 
+      function getTitle(tags) {
+          return l10n.tHtml('issues.fix.set_both_as_marked.title_use_marking', {
+            marking: tags['crossing:markings']
+          });
+      }
+
+
       return fixes;
     }
 
-    function makeCandidateFixes() {
+    function doMarkBothAsNode() {
+      return actionChangeTags(currentInfo.way.id, wayUpdateTags)(graph);
+    }
+
+
+    function doMarkBothAsWay() {
+      return actionChangeTags(currentInfo.node.id, nodeUpdateTags)(graph);
+    }
+
+
+
+    function doUnmarkNodeAsWay() {
+      return actionChangeTags(currentInfo.node.id, nodeDowngradeTags)(graph);
+    }
+
+    function makeMarkedUnmarkedFixes() {
       let fixes = [];
-      const graph = editor.staging.graph;  // I think we use staging graph in dynamic fixes?
+      const graph = editor.staging.graph;  // I think we use staging graph for dynamic fixes?
       const parentWay = graph.hasEntity(this.entityIds[1]);
 
-      fixes.push(
-        new ValidationFix({
-          icon: 'rapid-icon-crossing',
-          title: l10n.tHtml('issues.fix.use_crossing_tags_from_way.title'),
-          onClick: function () {
-            const graph = editor.staging.graph;
-            const [nodeID, wayID] = this.issue.entityIds;
-            const node = graph.hasEntity(nodeID);
-            const way = graph.hasEntity(wayID);
-            if (!node || !way) return;
-
-            const wayTags = way.tags;
-            let tags = {};
-
-            // At the very least, we need to make the node into a crossing node
-            tags.highway = 'crossing';
-            if (wayTags.crossing) {
-              tags.crossing = wayTags.crossing;
+      if (parentWay) {
+        fixes.push(
+          new ValidationFix({
+            icon: 'rapid-icon-crossing',
+            title: l10n.tHtml('issues.fix.set_both_as_marked.title'),
+            onClick:  () => {
+              const annotation = l10n.t('issues.fix.set_both_as_marked.annotation');
```

Comment:
I could be wrong, but I think in any of these onClick functions we still need the old defensive code to verify that the entities still exist - something like:
```js
const graph = editor.staging.graph;
const [nodeID, wayID] = this.issue.entityIds;
const node = graph.hasEntity(nodeID);
const way = graph.hasEntity(wayID);
if (!node || !way) return;
```

Because it might be possible for the user to change the thing that the issue was about and for the validator to lag a bit in validating that changed graph, and for the stale issue to still show up in a list somewhere and the user to be able to click the fix.

This is definitely one of those things where our validation code could be improved, but for now the defensive code doesn't hurt anything.

Reply:
Added defensive code, will update PR with changes today.